File: sdk/javascript/modelbridge.js
/**
 * Model Bridge JavaScript SDK - Enterprise Edition
 * Supports both Node.js and browser environments
 */

// TypeScript-style enums for JavaScript
const ModelProvider = {
    OPENAI: 'openai',
    ANTHROPIC: 'anthropic',
    GOOGLE: 'google',
    COHERE: 'cohere',
    GROQ: 'groq',
    DEEPSEEK: 'deepseek',
    MISTRAL: 'mistral',
    PERPLEXITY: 'perplexity',
    TOGETHER: 'together',
    HUGGINGFACE: 'huggingface',
    OLLAMA: 'ollama'
};

const WorkflowStatus = {
    DRAFT: 'draft',
    RUNNING: 'running',
    COMPLETED: 'completed',
    FAILED: 'failed',
    CANCELLED: 'cancelled'
};

/**
 * Model response data structure
 */
class ModelResponse {
    constructor(content, model, provider, usage = {}, metadata = {}) {
        this.content = content;
        this.model = model;
        this.provider = provider;
        this.usage = usage;
        this.metadata = metadata;
    }
}

/**
 * Workflow execution data structure
 */
class WorkflowExecution {
    constructor(id, workflowId, status, inputData, outputData, startedAt, completedAt = null, executionTimeMs = null, totalCost = null) {
        this.id = id;
        this.workflowId = workflowId;
        this.status = status;
        this.inputData = inputData;
        this.outputData = outputData;
        this.startedAt = startedAt;
        this.completedAt = completedAt;
        this.executionTimeMs = executionTimeMs;
        this.totalCost = totalCost;
    }
}

/**
 * Enterprise Model Bridge JavaScript SDK Client
 */
class ModelBridgeClient {
    constructor(apiKey, options = {}) {
        this.apiKey = apiKey;
        this.baseUrl = (options.baseUrl || 'http://localhost:8000').replace(/\/$/, '');
        this.organizationId = options.organizationId || null;
        this.timeout = options.timeout || 30000;
        this.maxRetries = options.maxRetries || 3;
        this.retryDelay = options.retryDelay || 1000;
        this.cache = new Map();
        
        // WebSocket support for real-time features
        this.ws = null;
        this.wsCallbacks = new Map();
    }

    /**
     * Make HTTP request with retry logic
     */
    async _makeRequest(method, endpoint, data = null, params = null) {
        const url = new URL(`${this.baseUrl}${endpoint}`);
        
        if (params) {
            Object.keys(params).forEach(key => {
                if (params[key] !== null && params[key] !== undefined) {
                    url.searchParams.append(key, params[key]);
                }
            });
        }

        const headers = {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json'
        };

        const config = {
            method,
            headers,
            timeout: this.timeout
        };

        if (data) {
            config.body = JSON.stringify(data);
        }

        for (let attempt = 0; attempt < this.maxRetries; attempt++) {
            try {
                const response = await fetch(url.toString(), config);
                
                if (response.ok) {
                    return await response.json();
                } else if (response.status === 401) {
                    throw new Error('Invalid API key');
                } else if (response.status === 403) {
                    throw new Error('Insufficient permissions');
                } else if (response.status === 429) {
                    if (attempt < this.maxRetries - 1) {
                        await this._sleep(this.retryDelay * Math.pow(2, attempt));
                        continue;
                    }
                    throw new Error('Rate limit exceeded');
                } else {
                    const errorData = await response.json();
                    throw new Error(`API error: ${errorData.detail || 'Unknown error'}`);
                }
            } catch (error) {
                if (attempt < this.maxRetries - 1) {
                    await this._sleep(this.retryDelay * Math.pow(2, attempt));
                    continue;
                }
                throw error;
            }
        }
        
        throw new Error('Max retries exceeded');
    }

    /**
     * Sleep utility for retry delays
     */
    _sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Send chat completion request
     */
    async chatCompletion(messages, options = {}) {
        const {
            model = 'gpt-3.5-turbo',
            provider = null,
            temperature = 0.7,
            maxTokens = null,
            stream = false,
            ...otherOptions
        } = options;

        const data = {
            messages,
            model,
            temperature,
            stream,
            ...otherOptions
        };

        if (provider) {
            data.provider = provider;
        }

        if (maxTokens) {
            data.max_tokens = maxTokens;
        }

        if (this.organizationId) {
            data.organization_id = this.organizationId;
        }

        const response = await this._makeRequest('POST', '/api/llm/chat', data);
        
        return new ModelResponse(
            response.choices[0].message.content,
            response.model,
            response.provider || 'unknown',
            response.usage || {},
            response.metadata || {}
        );
    }

    /**
     * Send completion request
     */
    async completion(prompt, options = {}) {
        const {
            model = 'gpt-3.5-turbo',
            provider = null,
            temperature = 0.7,
            maxTokens = null,
            ...otherOptions
        } = options;

        const data = {
            prompt,
            model,
            temperature,
            ...otherOptions
        };

        if (provider) {
            data.provider = provider;
        }

        if (maxTokens) {
            data.max_tokens = maxTokens;
        }

        if (this.organizationId) {
            data.organization_id = this.organizationId;
        }

        const response = await this._makeRequest('POST', '/api/llm/completion', data);
        
        return new ModelResponse(
            response.choices[0].text,
            response.model,
            response.provider || 'unknown',
            response.usage || {},
            response.metadata || {}
        );
    }

    /**
     * Get available models
     */
    async getModels() {
        const response = await this._makeRequest('GET', '/api/llm/models');
        return response.models || [];
    }

    /**
     * Get available providers
     */
    async getProviders() {
        const response = await this._makeRequest('GET', '/api/llm/providers');
        return response.providers || [];
    }

    /**
     * Create a new workflow
     */
    async createWorkflow(name, definition, description = null) {
        const data = {
            name,
            definition
        };

        if (description) {
            data.description = description;
        }

        if (this.organizationId) {
            data.organization_id = this.organizationId;
        }

        return await this._makeRequest('POST', '/api/workflows', data);
    }

    /**
     * Execute a workflow
     */
    async executeWorkflow(workflowId, inputData) {
        const data = {
            workflow_id: workflowId,
            input_data: inputData
        };

        if (this.organizationId) {
            data.organization_id = this.organizationId;
        }

        const response = await this._makeRequest('POST', '/api/workflows/execute', data);
        
        return new WorkflowExecution(
            response.id,
            response.workflow_id,
            response.status,
            response.input_data,
            response.output_data || {},
            new Date(response.started_at),
            response.completed_at ? new Date(response.completed_at) : null,
            response.execution_time_ms,
            response.total_cost
        );
    }

    /**
     * Get workflow execution status
     */
    async getWorkflowExecution(executionId) {
        const response = await this._makeRequest('GET', `/api/workflows/executions/${executionId}`);
        
        return new WorkflowExecution(
            response.id,
            response.workflow_id,
            response.status,
            response.input_data,
            response.output_data || {},
            new Date(response.started_at),
            response.completed_at ? new Date(response.completed_at) : null,
            response.execution_time_ms,
            response.total_cost
        );
    }

    /**
     * Get usage analytics
     */
    async getUsageAnalytics(options = {}) {
        const {
            startDate = null,
            endDate = null,
            groupBy = null
        } = options;

        const params = {};

        if (startDate) {
            params.start_date = startDate.toISOString();
        }

        if (endDate) {
            params.end_date = endDate.toISOString();
        }

        if (groupBy) {
            params.group_by = groupBy;
        }

        if (this.organizationId) {
            params.organization_id = this.organizationId;
        }

        return await this._makeRequest('GET', '/api/analytics/usage', null, params);
    }

    /**
     * Get cost analytics
     */
    async getCostAnalytics(options = {}) {
        const {
            startDate = null,
            endDate = null,
            costCenter = null
        } = options;

        const params = {};

        if (startDate) {
            params.start_date = startDate.toISOString();
        }

        if (endDate) {
            params.end_date = endDate.toISOString();
        }

        if (costCenter) {
            params.cost_center = costCenter;
        }

        if (this.organizationId) {
            params.organization_id = this.organizationId;
        }

        return await this._makeRequest('GET', '/api/analytics/costs', null, params);
    }

    /**
     * Get audit logs
     */
    async getAuditLogs(options = {}) {
        const {
            startDate = null,
            endDate = null,
            userId = null,
            action = null,
            limit = 100
        } = options;

        const params = { limit };

        if (startDate) {
            params.start_date = startDate.toISOString();
        }

        if (endDate) {
            params.end_date = endDate.toISOString();
        }

        if (userId) {
            params.user_id = userId;
        }

        if (action) {
            params.action = action;
        }

        if (this.organizationId) {
            params.organization_id = this.organizationId;
        }

        const response = await this._makeRequest('GET', '/api/audit/logs', null, params);
        return response.logs || [];
    }

    /**
     * Create a new role
     */
    async createRole(name, permissions, description = null) {
        const data = {
            name,
            permissions
        };

        if (description) {
            data.description = description;
        }

        if (this.organizationId) {
            data.organization_id = this.organizationId;
        }

        return await this._makeRequest('POST', '/api/rbac/roles', data);
    }

    /**
     * Assign role to user
     */
    async assignRole(userId, roleId) {
        const data = {
            user_id: userId,
            role_id: roleId
        };

        if (this.organizationId) {
            data.organization_id = this.organizationId;
        }

        return await this._makeRequest('POST', '/api/rbac/assignments', data);
    }

    /**
     * Get all roles
     */
    async getRoles() {
        const params = {};

        if (this.organizationId) {
            params.organization_id = this.organizationId;
        }

        const response = await this._makeRequest('GET', '/api/rbac/roles', null, params);
        return response.roles || [];
    }

    /**
     * Get all permissions
     */
    async getPermissions() {
        const response = await this._makeRequest('GET', '/api/rbac/permissions');
        return response.permissions || [];
    }

    /**
     * Connect to WebSocket for real-time updates
     */
    connectWebSocket() {
        if (typeof WebSocket === 'undefined') {
            throw new Error('WebSocket not available in this environment');
        }

        const wsUrl = this.baseUrl.replace('http', 'ws') + '/ws';
        this.ws = new WebSocket(wsUrl);

        this.ws.onopen = () => {
            console.log('WebSocket connected');
            // Send authentication
            this.ws.send(JSON.stringify({
                type: 'auth',
                api_key: this.apiKey
            }));
        };

        this.ws.onmessage = (event) => {
            try {
                const data = JSON.parse(event.data);
                this._handleWebSocketMessage(data);
            } catch (error) {
                console.error('Failed to parse WebSocket message:', error);
            }
        };

        this.ws.onclose = () => {
            console.log('WebSocket disconnected');
        };

        this.ws.onerror = (error) => {
            console.error('WebSocket error:', error);
        };
    }

    /**
     * Handle WebSocket messages
     */
    _handleWebSocketMessage(data) {
        const { type, payload } = data;
        
        if (this.wsCallbacks.has(type)) {
            this.wsCallbacks.get(type).forEach(callback => {
                try {
                    callback(payload);
                } catch (error) {
                    console.error('WebSocket callback error:', error);
                }
            });
        }
    }

    /**
     * Subscribe to WebSocket events
     */
    on(event, callback) {
        if (!this.wsCallbacks.has(event)) {
            this.wsCallbacks.set(event, []);
        }
        this.wsCallbacks.get(event).push(callback);
    }

    /**
     * Unsubscribe from WebSocket events
     */
    off(event, callback) {
        if (this.wsCallbacks.has(event)) {
            const callbacks = this.wsCallbacks.get(event);
            const index = callbacks.indexOf(callback);
            if (index > -1) {
                callbacks.splice(index, 1);
            }
        }
    }

    /**
     * Close WebSocket connection
     */
    disconnectWebSocket() {
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
    }
}

/**
 * Create a Model Bridge client
 */
function createClient(apiKey, options = {}) {
    return new ModelBridgeClient(apiKey, options);
}

/**
 * Quick chat completion function
 */
async function chatCompletion(apiKey, messages, options = {}) {
    const client = createClient(apiKey);
    return await client.chatCompletion(messages, options);
}

/**
 * Quick completion function
 */
async function completion(apiKey, prompt, options = {}) {
    const client = createClient(apiKey);
    return await client.completion(prompt, options);
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    // Node.js
    module.exports = {
        ModelBridgeClient,
        ModelResponse,
        WorkflowExecution,
        ModelProvider,
        WorkflowStatus,
        createClient,
        chatCompletion,
        completion
    };
} else if (typeof window !== 'undefined') {
    // Browser
    window.ModelBridge = {
        ModelBridgeClient,
        ModelResponse,
        WorkflowExecution,
        ModelProvider,
        WorkflowStatus,
        createClient,
        chatCompletion,
        completion
    };
} 

File: sdk/python/modelbridge.py
"""
Model Bridge Python SDK - Enterprise Edition
"""
import asyncio
import aiohttp
import json
import time
from typing import Dict, Any, Optional, List, Union
from datetime import datetime, timedelta
from dataclasses import dataclass
from enum import Enum


class ModelProvider(Enum):
    """Supported model providers"""
    OPENAI = "openai"
    ANTHROPIC = "anthropic"
    GOOGLE = "google"
    COHERE = "cohere"
    GROQ = "groq"
    DEEPSEEK = "deepseek"
    MISTRAL = "mistral"
    PERPLEXITY = "perplexity"
    TOGETHER = "together"
    HUGGINGFACE = "huggingface"
    OLLAMA = "ollama"


class WorkflowStatus(Enum):
    """Workflow execution status"""
    DRAFT = "draft"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


@dataclass
class ModelResponse:
    """Model response data"""
    content: str
    model: str
    provider: str
    usage: Dict[str, Any]
    metadata: Dict[str, Any]


@dataclass
class WorkflowExecution:
    """Workflow execution data"""
    id: str
    workflow_id: str
    status: WorkflowStatus
    input_data: Dict[str, Any]
    output_data: Dict[str, Any]
    started_at: datetime
    completed_at: Optional[datetime]
    execution_time_ms: Optional[int]
    total_cost: Optional[float]


class ModelBridgeClient:
    """Enterprise Model Bridge Python SDK Client"""
    
    def __init__(
        self,
        api_key: str,
        base_url: str = "http://localhost:8000",
        organization_id: Optional[str] = None,
        timeout: int = 30,
        max_retries: int = 3,
        retry_delay: float = 1.0
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.organization_id = organization_id
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.session = None
        self._cache = {}
    
    async def __aenter__(self):
        """Async context manager entry"""
        self.session = aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=self.timeout),
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            }
        )
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit"""
        if self.session:
            await self.session.close()
    
    async def _make_request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Make HTTP request with retry logic"""
        if not self.session:
            raise RuntimeError("Client not initialized. Use async context manager.")
        
        url = f"{self.base_url}{endpoint}"
        
        for attempt in range(self.max_retries):
            try:
                async with self.session.request(
                    method=method,
                    url=url,
                    json=data,
                    params=params
                ) as response:
                    if response.status == 200:
                        return await response.json()
                    elif response.status == 401:
                        raise AuthenticationError("Invalid API key")
                    elif response.status == 403:
                        raise PermissionError("Insufficient permissions")
                    elif response.status == 429:
                        if attempt < self.max_retries - 1:
                            await asyncio.sleep(self.retry_delay * (2 ** attempt))
                            continue
                        raise RateLimitError("Rate limit exceeded")
                    else:
                        error_data = await response.json()
                        raise APIError(f"API error: {error_data.get('detail', 'Unknown error')}")
                        
            except aiohttp.ClientError as e:
                if attempt < self.max_retries - 1:
                    await asyncio.sleep(self.retry_delay * (2 ** attempt))
                    continue
                raise ConnectionError(f"Connection error: {str(e)}")
        
        raise APIError("Max retries exceeded")
    
    async def chat_completion(
        self,
        messages: List[Dict[str, str]],
        model: str = "gpt-3.5-turbo",
        provider: Optional[ModelProvider] = None,
        temperature: float = 0.7,
        max_tokens: Optional[int] = None,
        stream: bool = False,
        **kwargs
    ) -> ModelResponse:
        """Send chat completion request"""
        data = {
            "messages": messages,
            "model": model,
            "temperature": temperature,
            "stream": stream,
            **kwargs
        }
        
        if provider:
            data["provider"] = provider.value
        
        if max_tokens:
            data["max_tokens"] = max_tokens
        
        if self.organization_id:
            data["organization_id"] = self.organization_id
        
        response = await self._make_request("POST", "/api/llm/chat", data=data)
        
        return ModelResponse(
            content=response["choices"][0]["message"]["content"],
            model=response["model"],
            provider=response.get("provider", "unknown"),
            usage=response.get("usage", {}),
            metadata=response.get("metadata", {})
        )
    
    async def completion(
        self,
        prompt: str,
        model: str = "gpt-3.5-turbo",
        provider: Optional[ModelProvider] = None,
        temperature: float = 0.7,
        max_tokens: Optional[int] = None,
        **kwargs
    ) -> ModelResponse:
        """Send completion request"""
        data = {
            "prompt": prompt,
            "model": model,
            "temperature": temperature,
            **kwargs
        }
        
        if provider:
            data["provider"] = provider.value
        
        if max_tokens:
            data["max_tokens"] = max_tokens
        
        if self.organization_id:
            data["organization_id"] = self.organization_id
        
        response = await self._make_request("POST", "/api/llm/completion", data=data)
        
        return ModelResponse(
            content=response["choices"][0]["text"],
            model=response["model"],
            provider=response.get("provider", "unknown"),
            usage=response.get("usage", {}),
            metadata=response.get("metadata", {})
        )
    
    async def get_models(self) -> List[Dict[str, Any]]:
        """Get available models"""
        response = await self._make_request("GET", "/api/llm/models")
        return response.get("models", [])
    
    async def get_providers(self) -> List[Dict[str, Any]]:
        """Get available providers"""
        response = await self._make_request("GET", "/api/llm/providers")
        return response.get("providers", [])
    
    async def create_workflow(
        self,
        name: str,
        definition: Dict[str, Any],
        description: Optional[str] = None
    ) -> Dict[str, Any]:
        """Create a new workflow"""
        data = {
            "name": name,
            "definition": definition
        }
        
        if description:
            data["description"] = description
        
        if self.organization_id:
            data["organization_id"] = self.organization_id
        
        response = await self._make_request("POST", "/api/workflows", data=data)
        return response
    
    async def execute_workflow(
        self,
        workflow_id: str,
        input_data: Dict[str, Any]
    ) -> WorkflowExecution:
        """Execute a workflow"""
        data = {
            "workflow_id": workflow_id,
            "input_data": input_data
        }
        
        if self.organization_id:
            data["organization_id"] = self.organization_id
        
        response = await self._make_request("POST", "/api/workflows/execute", data=data)
        
        return WorkflowExecution(
            id=response["id"],
            workflow_id=response["workflow_id"],
            status=WorkflowStatus(response["status"]),
            input_data=response["input_data"],
            output_data=response.get("output_data", {}),
            started_at=datetime.fromisoformat(response["started_at"]),
            completed_at=datetime.fromisoformat(response["completed_at"]) if response.get("completed_at") else None,
            execution_time_ms=response.get("execution_time_ms"),
            total_cost=response.get("total_cost")
        )
    
    async def get_workflow_execution(self, execution_id: str) -> WorkflowExecution:
        """Get workflow execution status"""
        response = await self._make_request("GET", f"/api/workflows/executions/{execution_id}")
        
        return WorkflowExecution(
            id=response["id"],
            workflow_id=response["workflow_id"],
            status=WorkflowStatus(response["status"]),
            input_data=response["input_data"],
            output_data=response.get("output_data", {}),
            started_at=datetime.fromisoformat(response["started_at"]),
            completed_at=datetime.fromisoformat(response["completed_at"]) if response.get("completed_at") else None,
            execution_time_ms=response.get("execution_time_ms"),
            total_cost=response.get("total_cost")
        )
    
    async def get_usage_analytics(
        self,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        group_by: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get usage analytics"""
        params = {}
        
        if start_date:
            params["start_date"] = start_date.isoformat()
        
        if end_date:
            params["end_date"] = end_date.isoformat()
        
        if group_by:
            params["group_by"] = group_by
        
        if self.organization_id:
            params["organization_id"] = self.organization_id
        
        response = await self._make_request("GET", "/api/analytics/usage", params=params)
        return response
    
    async def get_cost_analytics(
        self,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        cost_center: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get cost analytics"""
        params = {}
        
        if start_date:
            params["start_date"] = start_date.isoformat()
        
        if end_date:
            params["end_date"] = end_date.isoformat()
        
        if cost_center:
            params["cost_center"] = cost_center
        
        if self.organization_id:
            params["organization_id"] = self.organization_id
        
        response = await self._make_request("GET", "/api/analytics/costs", params=params)
        return response
    
    async def get_audit_logs(
        self,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        user_id: Optional[str] = None,
        action: Optional[str] = None,
        limit: int = 100
    ) -> List[Dict[str, Any]]:
        """Get audit logs"""
        params = {"limit": limit}
        
        if start_date:
            params["start_date"] = start_date.isoformat()
        
        if end_date:
            params["end_date"] = end_date.isoformat()
        
        if user_id:
            params["user_id"] = user_id
        
        if action:
            params["action"] = action
        
        if self.organization_id:
            params["organization_id"] = self.organization_id
        
        response = await self._make_request("GET", "/api/audit/logs", params=params)
        return response.get("logs", [])
    
    async def create_role(
        self,
        name: str,
        permissions: List[str],
        description: Optional[str] = None
    ) -> Dict[str, Any]:
        """Create a new role"""
        data = {
            "name": name,
            "permissions": permissions
        }
        
        if description:
            data["description"] = description
        
        if self.organization_id:
            data["organization_id"] = self.organization_id
        
        response = await self._make_request("POST", "/api/rbac/roles", data=data)
        return response
    
    async def assign_role(
        self,
        user_id: str,
        role_id: str
    ) -> Dict[str, Any]:
        """Assign role to user"""
        data = {
            "user_id": user_id,
            "role_id": role_id
        }
        
        if self.organization_id:
            data["organization_id"] = self.organization_id
        
        response = await self._make_request("POST", "/api/rbac/assignments", data=data)
        return response
    
    async def get_roles(self) -> List[Dict[str, Any]]:
        """Get all roles"""
        params = {}
        
        if self.organization_id:
            params["organization_id"] = self.organization_id
        
        response = await self._make_request("GET", "/api/rbac/roles", params=params)
        return response.get("roles", [])
    
    async def get_permissions(self) -> List[Dict[str, Any]]:
        """Get all permissions"""
        response = await self._make_request("GET", "/api/rbac/permissions")
        return response.get("permissions", [])


# Exception classes
class ModelBridgeError(Exception):
    """Base exception for Model Bridge SDK"""
    pass


class AuthenticationError(ModelBridgeError):
    """Authentication error"""
    pass


class PermissionError(ModelBridgeError):
    """Permission error"""
    pass


class RateLimitError(ModelBridgeError):
    """Rate limit error"""
    pass


class APIError(ModelBridgeError):
    """API error"""
    pass


class ConnectionError(ModelBridgeError):
    """Connection error"""
    pass


# Convenience functions for synchronous usage
def create_client(
    api_key: str,
    base_url: str = "http://localhost:8000",
    organization_id: Optional[str] = None,
    timeout: int = 30,
    max_retries: int = 3,
    retry_delay: float = 1.0
) -> ModelBridgeClient:
    """Create a Model Bridge client"""
    return ModelBridgeClient(
        api_key=api_key,
        base_url=base_url,
        organization_id=organization_id,
        timeout=timeout,
        max_retries=max_retries,
        retry_delay=retry_delay
    )


async def chat_completion(
    api_key: str,
    messages: List[Dict[str, str]],
    model: str = "gpt-3.5-turbo",
    provider: Optional[ModelProvider] = None,
    **kwargs
) -> ModelResponse:
    """Quick chat completion"""
    async with create_client(api_key) as client:
        return await client.chat_completion(messages, model, provider, **kwargs)


async def completion(
    api_key: str,
    prompt: str,
    model: str = "gpt-3.5-turbo",
    provider: Optional[ModelProvider] = None,
    **kwargs
) -> ModelResponse:
    """Quick completion"""
    async with create_client(api_key) as client:
        return await client.completion(prompt, model, provider, **kwargs) 